common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineEntitlementsManager.kt

```diff
@@ -1,15 +1,64 @@
 package com.revenuecat.purchases.common.offlineentitlements
 
+import com.revenuecat.purchases.CustomerInfo
+import com.revenuecat.purchases.PurchasesError
 import com.revenuecat.purchases.common.Backend
 import com.revenuecat.purchases.common.caching.DeviceCache
 import com.revenuecat.purchases.common.debugLog
 import com.revenuecat.purchases.common.errorLog
+import com.revenuecat.purchases.common.warnLog
 import com.revenuecat.purchases.strings.OfflineEntitlementsStrings
 
 class OfflineEntitlementsManager(
     private val backend: Backend,
+    private val offlineCustomerInfoCalculator: OfflineCustomerInfoCalculator,
     private val deviceCache: DeviceCache
 ) {
+    // We cache the offline customer info in memory, so it's not persisted.
+    val offlineCustomerInfo: CustomerInfo?
+        get() = _offlineCustomerInfo
+    // TODO make sure we handle concurrency
+    private var _offlineCustomerInfo: CustomerInfo? = null
+
+    fun resetOfflineCustomerInfoCache() {
+        if (_offlineCustomerInfo != null) {
+            // TODO improve logs
+            warnLog("Resetting offline customer info cache")
+        }
+        _offlineCustomerInfo = null
+    }
+
+    fun shouldCalculateOfflineCustomerInfoInGetCustomerInfoRequest(
+        isServerError: Boolean,
+        appUserId: String
+    ): Boolean {
+        return isServerError && deviceCache.getCachedCustomerInfo(appUserId) == null
```

Comment:
this is a good catch - without this, if we just rely on the error code, we'll be sending offline customer info to the delegate needlessly, since we already have a cached value and it might be up to date as far as we know

Reply:
Right, this is part of the triggers in the doc. This makes it so we don't enable offline entitlements if there is a cached value already. But if the user posted a receipt first, enabling offline entitlements, we would still return that over the cached customer info